File: solutions/haiku-1/haiku/perl/dijkstra.pl
#!/usr/bin/perl
use strict;
use warnings;

# Read N and M
my ($N, $M) = split(' ', <STDIN>);

# Build adjacency list
my %graph;
for my $i (0 .. $N - 1) {
    $graph{$i} = [];
}

# Read edges
for (1 .. $M) {
    my ($u, $v, $w) = split(' ', <STDIN>);
    push @{$graph{$u}}, [$v, $w];
    push @{$graph{$v}}, [$u, $w];
}

# Read s and t
my ($s, $t) = split(' ', <STDIN>);

# Special case: s == t
if ($s == $t) {
    print "0\n";
    exit;
}

# Dijkstra's algorithm
my @dist = (1e9) x $N;
$dist[$s] = 0;

my %visited;

# N iterations to relax all edges
for (1 .. $N) {
    # Find unvisited node with minimum distance
    my $u = -1;
    my $minDist = 1e9;

    for my $i (0 .. $N - 1) {
        if (!$visited{$i} && $dist[$i] < $minDist) {
            $minDist = $dist[$i];
            $u = $i;
        }
    }

    # No more reachable nodes
    last if $u == -1 || $minDist == 1e9;

    $visited{$u} = 1;

    # Relax edges from u
    for my $edge (@{$graph{$u}}) {
        my ($v, $w) = @$edge;
        if ($dist[$u] + $w < $dist[$v]) {
            $dist[$v] = $dist[$u] + $w;
        }
    }
}

# Output result
if ($dist[$t] == 1e9) {
    print "-1\n";
} else {
    print $dist[$t] . "\n";
}

File: solutions/haiku-1/haiku/typescript/dijkstra.ts
import * as readline from "readline";

interface Edge {
  to: number;
  weight: number;
}

async function main() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
  });

  const lines: string[] = [];
  for await (const line of rl) {
    lines.push(line.trim());
  }

  // Parse input
  const [n, m] = lines[0].split(" ").map(Number);

  // Build adjacency list
  const graph: Edge[][] = Array.from({ length: n }, () => []);

  for (let i = 1; i <= m; i++) {
    const [u, v, w] = lines[i].split(" ").map(Number);
    graph[u].push({ to: v, weight: w });
    graph[v].push({ to: u, weight: w });
  }

  const [s, t] = lines[m + 1].split(" ").map(Number);

  // Special case: start equals target
  if (s === t) {
    console.log(0);
    return;
  }

  // Dijkstra's algorithm
  const dist = Array(n).fill(Infinity);
  dist[s] = 0;

  const visited = Array(n).fill(false);
  const pq: [number, number][] = [[0, s]]; // [distance, node]

  while (pq.length > 0) {
    // Find the unvisited node with minimum distance
    let minIdx = 0;
    for (let i = 1; i < pq.length; i++) {
      if (pq[i][0] < pq[minIdx][0]) {
        minIdx = i;
      }
    }

    const [d, u] = pq[minIdx];
    pq.splice(minIdx, 1);

    if (visited[u]) continue;
    visited[u] = true;

    if (d > dist[u]) continue;

    // Early termination if we reached the target
    if (u === t) {
      console.log(dist[t]);
      return;
    }

    for (const edge of graph[u]) {
      const v = edge.to;
      const w = edge.weight;

      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w;
        pq.push([dist[v], v]);
      }
    }
  }

  // Check if target is reachable
  if (dist[t] === Infinity) {
    console.log(-1);
  } else {
    console.log(dist[t]);
  }
}

main().catch(console.error);
